File: src/services/terminal.ts
import { api } from "./api";
import { terminalCache } from "../core/performance";
import type {
  TerminalTitleChanged,
  TerminalExited,
  CreateTerminalResponse,
  WriteTerminalRequest,
  ResizeTerminalRequest,
  TerminalInfo,
  TerminalData,
  TerminalLatency,
} from "../types/panel";

let outputUnlisten: (() => void) | null = null;
let titleUnlisten: (() => void) | null = null;
let exitUnlisten: (() => void) | null = null;
let latencyUnlisten: (() => void) | null = null;

/**
 * Create a new local terminal
 */
export async function createTerminal(
  shell?: string,
  workingDir?: string,
  title?: string,
): Promise<CreateTerminalResponse> {
  try {
    return await api.call<CreateTerminalResponse>("create_terminal", {
      shell,
      workingDir,
      title,
    });
  } catch (error) {
    throw error;
  }
}

/**
 * Create a new local terminal (alias for createTerminal)
 */
export const createLocalTerminal = createTerminal;

/**
 * Create a new SSH terminal using profile ID
 */
export async function createSSHTerminal(
  profileId: string,
): Promise<CreateTerminalResponse> {
  try {
    return await api.call<CreateTerminalResponse>("create_ssh_terminal", {
      profileId,
    });
  } catch (error) {
    throw error;
  }
}

/**
 * Create a new SSH terminal from SSH config host
 */
export async function createSSHConfigTerminal(
  hostName: string,
  title?: string,
  password?: string,
): Promise<CreateTerminalResponse> {
  try {
    return await api.call<CreateTerminalResponse>(
      "create_ssh_config_terminal",
      {
        hostName,
        title,
        password,
      },
    );
  } catch (error) {
    throw error;
  }
}

/**
 * Write data to a terminal
 */
export async function writeToTerminal(
  request: WriteTerminalRequest,
): Promise<void> {
  try {
    return await api.call<void>("write_to_terminal", request);
  } catch (error) {
    throw error;
  }
}

/**
 * Write data to multiple terminals in batch
 */
export async function writeBatchToTerminal(
  requests: WriteTerminalRequest[],
): Promise<void> {
  try {
    return await api.call<void>("write_batch_to_terminal", {
      requests,
    });
  } catch (error) {
    console.error("Failed to write batch to terminals:", error);
    throw error;
  }
}

/**
 * Resize a terminal
 */
export async function resizeTerminal(
  request: ResizeTerminalRequest,
): Promise<void> {
  try {
    return await api.call<void>("resize_terminal", request);
  } catch (error) {
    console.error("Failed to resize terminal:", error);
    throw error;
  }
}

/**
 * Close a specific terminal
 */
export async function closeTerminal(terminalId: string): Promise<void> {
  if (!terminalId || terminalId.trim() === "") {
    throw new Error("Terminal ID is required and cannot be empty");
  }

  try {
    const result = await api.call<void>("close_terminal", {
      terminalId,
    });

    terminalCache.invalidateTerminal(terminalId);

    return result;
  } catch (error) {
    console.error("Failed to close terminal:", error);
    throw error;
  }
}

/**
 * Get information about a specific terminal (cached)
 */
export async function getTerminalInfo(
  terminalId: string,
): Promise<TerminalInfo> {
  try {
    return await terminalCache.getTerminalInfo(terminalId);
  } catch (error) {
    console.error("Failed to get terminal info:", error);
    throw error;
  }
}

/**
 * List all active terminals (cached)
 */
export async function listTerminals(): Promise<TerminalInfo[]> {
  try {
    return await terminalCache.getTerminalList();
  } catch (error) {
    console.error("Failed to list terminals:", error);
    throw error;
  }
}

/**
 * Get user@hostname for terminal title
 */
export async function getUserHostname(): Promise<string> {
  try {
    return await api.callRaw<string>("get_user_hostname");
  } catch (error) {
    console.error("Failed to get user hostname:", error);
    return "user@hostname"; // fallback
  }
}

/**
 * Listen to terminal output events
 */
export async function listenToTerminalOutput(
  callback: (data: TerminalData) => void,
): Promise<() => void> {
  outputUnlisten = await api.listen<TerminalData>("terminal-output", callback);
  return outputUnlisten;
}

/**
 * Listen to terminal title change events
 */
export async function listenToTerminalTitleChanged(
  callback: (data: TerminalTitleChanged) => void,
): Promise<() => void> {
  titleUnlisten = await api.listen<TerminalTitleChanged>(
    "terminal-title-changed",
    callback,
  );
  return titleUnlisten;
}

/**
 * Listen to terminal exit events
 */
export async function listenToTerminalExit(
  callback: (data: TerminalExited) => void,
): Promise<() => void> {
  exitUnlisten = await api.listen<TerminalExited>("terminal-exited", callback);
  return exitUnlisten;
}

/**
 * Listen to terminal latency events
 */
export async function listenToTerminalLatency(
  callback: (data: TerminalLatency) => void,
): Promise<() => void> {
  latencyUnlisten = await api.listen<TerminalLatency>(
    "terminal-latency",
    callback,
  );
  return latencyUnlisten;
}

/**
 * Cleanup all terminal listeners
 */
export const cleanupTerminalListeners = (): void => {
  if (outputUnlisten) {
    outputUnlisten();
    outputUnlisten = null;
  }
  if (titleUnlisten) {
    titleUnlisten();
    titleUnlisten = null;
  }
  if (exitUnlisten) {
    exitUnlisten();
    exitUnlisten = null;
  }
  if (latencyUnlisten) {
    latencyUnlisten();
    latencyUnlisten = null;
  }
};

File: src-tauri/src/commands/terminal.rs
use crate::error::AppError;
use crate::models::terminal::{
    CloseTerminalRequest, CreateLocalTerminalRequest, CreateSshConfigTerminalRequest,
    CreateSshTerminalRequest, CreateTerminalRequest, CreateTerminalResponse,
    GetTerminalInfoRequest, LocalConfig, ResizeTerminalRequest, TerminalConfig, TerminalInfo,
    TerminalType, WriteBatchTerminalRequest, WriteTerminalRequest,
};
use crate::state::AppState;
use tauri::{AppHandle, State};

/// Create a new local terminal
#[tauri::command]
pub async fn create_terminal(
    request: CreateLocalTerminalRequest,
    app_state: State<'_, AppState>,
    app_handle: AppHandle,
) -> Result<CreateTerminalResponse, AppError> {
    let config = TerminalConfig {
        terminal_type: TerminalType::Local,
        local_config: Some(LocalConfig {
            shell: request.shell,
            working_dir: request.working_dir,
            env_vars: None,
        }),
        ssh_profile_id: None,
        ssh_config_host: None,
        ssh_config_password: None,
    };

    let create_request = CreateTerminalRequest {
        config,
        title: request.title,
    };
    app_state
        .terminal_manager
        .create_terminal(create_request, Some(app_handle))
        .await
}

/// Create a new SSH terminal using profile ID
#[tauri::command]
pub async fn create_ssh_terminal(
    request: CreateSshTerminalRequest,
    app_state: State<'_, AppState>,
    app_handle: AppHandle,
) -> Result<CreateTerminalResponse, AppError> {
    let ssh_profile = app_state
        .ssh_service
        .get_ssh_profile(&request.profile_id)
        .await
        .map_err(|e| AppError::Database(e.to_string()))?;

    let config = TerminalConfig {
        terminal_type: TerminalType::SSH,
        local_config: None,
        ssh_profile_id: Some(request.profile_id),
        ssh_config_host: None,
        ssh_config_password: None,
    };

    let terminal_request = CreateTerminalRequest {
        config,
        title: Some(ssh_profile.display_name()),
    };

    app_state
        .terminal_manager
        .create_terminal(terminal_request, Some(app_handle))
        .await
}

/// Create a new SSH terminal from SSH config host
#[tauri::command]
pub async fn create_ssh_config_terminal(
    request: CreateSshConfigTerminalRequest,
    app_state: State<'_, AppState>,
    app_handle: AppHandle,
) -> Result<CreateTerminalResponse, AppError> {
    use crate::services::ssh_config_parser::parse_ssh_config;

    let hosts = parse_ssh_config(None)
        .await
        .map_err(|e| AppError::Config(format!("Failed to parse SSH config: {}", e)))?;

    let host = hosts
        .iter()
        .find(|h| h.name == request.host_name)
        .ok_or_else(|| {
            AppError::Config(format!(
                "Host '{}' not found in SSH config",
                request.host_name
            ))
        })?;

    // Check if password is required but not provided
    if host.requires_password() && request.password.is_none() {
        return Err(AppError::authentication_failed(
            "Password is required for this host",
        ));
    }

    let config = TerminalConfig {
        terminal_type: TerminalType::SSHConfig,
        local_config: None,
        ssh_profile_id: None,
        ssh_config_host: Some(host.clone()),
        ssh_config_password: request.password,
    };

    let title = request.title.unwrap_or_else(|| {
        format!(
            "{}@{}",
            host.user.as_deref().unwrap_or("user"),
            host.hostname
        )
    });

    let terminal_request = CreateTerminalRequest {
        config,
        title: Some(title),
    };

    app_state
        .terminal_manager
        .create_terminal(terminal_request, Some(app_handle))
        .await
}

/// Write data to a terminal
#[tauri::command]
pub async fn write_to_terminal(
    request: WriteTerminalRequest,
    app_state: State<'_, AppState>,
) -> Result<(), AppError> {
    app_state.terminal_manager.write_to_terminal(request).await
}

/// Write data to multiple terminals in batch
#[tauri::command]
pub async fn write_batch_to_terminal(
    request: WriteBatchTerminalRequest,
    app_state: State<'_, AppState>,
) -> Result<(), AppError> {
    for write_request in request.requests {
        app_state
            .terminal_manager
            .write_to_terminal(write_request)
            .await?;
    }
    Ok(())
}

/// Resize a terminal
#[tauri::command]
pub async fn resize_terminal(
    request: ResizeTerminalRequest,
    app_state: State<'_, AppState>,
) -> Result<(), AppError> {
    app_state.terminal_manager.resize_terminal(request).await
}

/// Close a specific terminal
#[tauri::command]
pub async fn close_terminal(
    request: CloseTerminalRequest,
    app_state: State<'_, AppState>,
) -> Result<(), AppError> {
    app_state
        .terminal_manager
        .close_terminal(request.terminal_id)
        .await
}

/// Get information about a specific terminal
#[tauri::command]
pub async fn get_terminal_info(
    request: GetTerminalInfoRequest,
    app_state: State<'_, AppState>,
) -> Result<TerminalInfo, AppError> {
    app_state
        .terminal_manager
        .get_terminal_info(request.terminal_id)
        .await
}

/// List all active terminals
#[tauri::command]
pub async fn list_terminals(app_state: State<'_, AppState>) -> Result<Vec<TerminalInfo>, AppError> {
    app_state.terminal_manager.list_terminals().await
}
